lib/grizzly/trace.ex

```diff
@@ -133,41 +129,60 @@ defmodule Grizzly.Trace do
   def init(opts) do
     size = Keyword.get(opts, :size, @default_size)
     record_keepalives = Keyword.get(opts, :record_keepalives, true)
-    {:ok, %{buffer: CircularBuffer.new(size), record_keepalives: record_keepalives}}
+    {:ok, %{buffers: %{}, size: size, record_keepalives: record_keepalives}}
   end
 
   @impl GenServer
-  def handle_cast({:log, <<0x23, 0x03, _ack_flag>>, _opts}, %{record_keepalives: false} = state) do
+  def handle_cast(
+        {:log, _src, _dest, <<0x23, 0x03, _ack_flag>>},
+        %{record_keepalives: false} = state
+      ) do
     {:noreply, state}
   end
 
-  def handle_cast({:log, binary, opts}, state) do
-    record = Record.new(binary, opts)
+  def handle_cast({:log, src, dest, binary}, state) do
+    record = Record.new(src, dest, binary)
+
+    remote_node = Record.remote_node(record)
+
+    state =
+      update_in(state.buffers[remote_node], fn
+        nil -> state.size |> CircularBuffer.new() |> CircularBuffer.insert(record)
+        buf -> CircularBuffer.insert(buf, record)
+      end)
 
-    {:noreply, %{state | buffer: CircularBuffer.insert(state.buffer, record)}}
+    {:noreply, state}
   end
 
   @impl GenServer
-  def handle_call(:clear, _from, %{buffer: %CircularBuffer{max_size: size}} = state) do
-    {:reply, :ok, %{state | buffer: CircularBuffer.new(size)}}
+  def handle_call(:clear, _from, state) do
+    {:reply, :ok, %{state | buffers: %{}}}
   end
 
-  def handle_call(:list, _from, %{buffer: buffer} = state) do
-    {:reply, CircularBuffer.to_list(buffer), state}
+  def handle_call(:list, _from, state) do
+    records =
+      for buffer <- Map.values(state.buffers),
+          record <- buffer,
+          do: record
+
+    {:reply, Enum.sort_by(records, & &1.timestamp, {:asc, Time}), state}
   end
 
   def handle_call({:record_keepalives, enabled?}, _from, state) do
     {:reply, :ok, %{state | record_keepalives: enabled?}}
   end
 
-  def handle_call({:resize, size}, _from, %{buffer: buffer} = state) do
-    new_buffer =
-      buffer
-      |> CircularBuffer.to_list()
-      |> Enum.reduce(CircularBuffer.new(size), fn record, buffer ->
-        CircularBuffer.insert(buffer, record)
+  def handle_call({:resize, new_size}, _from, %{buffers: buffers} = state) do
+    new_buffers =
+      Map.new(buffers, fn {node_id, buffer} ->
+        resized_buffer =
+          Enum.reduce(buffer, CircularBuffer.new(new_size), fn record, new_buffer ->
+            CircularBuffer.insert(new_buffer, record)
+          end)
+
+        {node_id, resized_buffer}
       end)
```

Comment:
TIL [that `CircularBuffer` implements `Collectable`](https://github.com/elixir-toniq/circular_buffer/blob/main/lib/circular_buffer.ex#L138-L139).

So, tiny refactor idea:

```suggestion
    new_buffers =
      Map.new(buffers, fn {node_id, buffer} ->
        resized_buffer = Enum.into(buffer, CircularBuffer.new(new_size))
        {node_id, resized_buffer}
      end)
```